### PR TITLE
Added leaflet node plugin which allows users to add an open street map node link

### PR DESCRIPTION
**Pull request by Goldmaster**
This merge request is to add a simple JavaScript plugin that allows users to paste in an open street maps node link and display info about the node on a map. This was created due to myself rasing an issue https://github.com/Leaflet/Leaflet/issues/9151

### CODE REVIEW

File: docs/_plugins/data-providers/leaflet-node.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+---
+name: Leaflet node
+category: Data Providers
+repo: [Gitlab](https://gitlab.com/Goldmaster/leaflet-node)
+author: Goldmaster
+author-url: https://GoldmastersITDM.co.uk
+demo: https://goldmaster.gitlab.io/leaflet-node
```
Comment (Falke-Design) [inline docs/_plugins/data-providers/leaflet-node.md:7]: Not working

File: docs/_plugins/data-providers/leaflet-node.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+---
+name: Leaflet node
+category: Data Providers
+repo: [Gitlab](https://gitlab.com/Goldmaster/leaflet-node)
```
Comment (Falke-Design) [inline docs/_plugins/data-providers/leaflet-node.md:4]:
```suggestion
repo: https://gitlab.com/Goldmaster/leaflet-node
```

**Review by Falke-Design — CHANGES_REQUESTED** [review]

File: docs/_plugins/data-providers/leaflet-node.md (hunk 1 above)
Comment (Goldmaster) [reply docs/_plugins/data-providers/leaflet-node.md:7]: Hi there, I'm afraid I am unsure why. I did follow the guide on getting the demo to work as it runs on gitlab pages.

Despite following the correct info, the demo fails to show.

I do have a real world demo example linked as well.

I will do some further research on this issue.

Comment (Goldmaster) [reply docs/_plugins/data-providers/leaflet-node.md:7]: Have a check now, got the demo to work

File: docs/_plugins/data-providers/leaflet-node.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+---
+name: Leaflet node
+category: Data Providers
+repo: https://gitlab.com/Goldmaster/leaflet-node
+author: Goldmaster
+author-url: https://GoldmastersITDM.co.uk
```
Comment (jonkoops) [inline docs/_plugins/data-providers/leaflet-node.md:6]: Let's change this to the final URL we get redirected to.
```suggestion
author-url: https://goldmastersitdm.co.uk/
```

**Review by jonkoops — CHANGES_REQUESTED** [review]

**Comment by Goldmaster** [comment]
i have fixed the website link. i do also see an update branch button, as i believe the fork and edited copy may not be as up-to-date as. do i need to click it? Many thanks

**Comment by bonacciniWd** [comment]
Could you start the lime props using useStates and useEffect, then use the prop to render the default system.

**Comment by Falke-Design** [comment]
Sorry I didn't see it on the first review, but I can't merge this because it is no plugin, it is just a scrapping script for OSM.

**Comment by Goldmaster** [comment]
> Sorry I didn't see it on the first review, but I can't merge this because it is no plugin, it is just a scrapping script for OSM.

Ok, the basic idea is to be able to show info about a particular node similar to Google maps. The user would put in a osm node link and on a map would show a point with info such as address, phone, website ect.

**Comment by Falke-Design** [comment]
Sure the idea is good, if you provide it in a look and feel like a Leaflet Plugin, I will add it.

So what does this mean:
* Don't create the map in your function, a plugin is addtionally and normally not creating the map
* The arguments should be passed into a function as options and not as global variable

Like this:

```
L.NodeMarker  = L.Marker.extend({
    initialize(nodeId, options) {
		L.Util.setOptions(this, options);
		this._nodeId = nodeId;
    },
    onAdd: function(map) {
        const osmNodeLink = 'https://www.openstreetmap.org/node/'+this._nodeId;
        this.extractDataFromOSMNodeLink(osmNodeLink)
            .then(data => {
                  this._nodeData = data;
                  this._latlng = L.latLng(data.latitude, data.longitude)
                  L.Marker.prototype.onAdd.call(this, map)
                  this._bindPopupData();
			      // because of async promise, we fire a new event
                  this.fire('nodemarker:add', {layer: this})
            })
    },
    extractDataFromOSMNodeLink(link) {
            // Regular expression to match the node ID
            const nodeIDMatch = link.match(/\/node\/(\d+)/);

            if (nodeIDMatch && nodeIDMatch.length === 2) {
                const nodeID = nodeIDMatch[1];
                return fetch(`https://api.openstreetmap.org/api/0.6/node/${nodeID}`)
                    .then(response => response.text())
                    .then(data => {
                        const parser = new DOMParser();
                        const xmlDoc = parser.parseFromString(data, "text/xml");
                        const result = { latitude: xmlDoc.querySelector("node").getAttribute("lat"), longitude: xmlDoc.querySelector("node").getAttribute("lon") };

                        this.options.dataVariables.forEach(variable => {
                            const value = xmlDoc.querySelector(`node > tag[k='${variable.name}']`);
                            if (value) {
                                result[variable.name] = variable.name === 'website' ? this.formatWebsite(value.getAttribute("v")) : variable.name === 'phone' ? this.formatPhoneNumber(value.getAttribute("v")) : value.getAttribute("v");
                            }
                        });

                        return result;
                    });
            } else {
                return Promise.reject("Invalid or unsupported OSM link format");
            }
        },
        _bindPopupData(){
                const popupContent = `
                  <div style="background-color: white; padding: 10px;">
                    ${this.options.dataVariables.map(variable => this._nodeData[variable.name] ? `<p><strong>${variable.label}:</strong> ${this._nodeData[variable.name]}</p>` : '').join('')}
                  </div>
                `;

               this.bindPopup(popupContent);
        },
// Function to format the website URL
        formatWebsite(website) {
            if (website) {
                website = website.replace(/^(https?:\/\/)?(www\.)?/, '');
                return `<a href="https://${website}" target="_blank">${website}</a>`;
            }
            return '';
        },

        // Function to format the phone number with "tel:" prefix
        formatPhoneNumber(phone) {
            if (phone) {
                return `<a href="tel:${phone}">${phone}</a>`;
            }
            return '';
        }
});
```
```
var nodeMarker = new L.NodeMarker('3060887211', {
        dataVariables: [
            { name: "alt_name", label: "Name of Place" },
            { name: "addr:housename", label: "House Name" },
            { name: "addr:housenumber", label: "House Number" },
            { name: "addr:postcode", label: "Postcode" },
            { name: "addr:street", label: "Street" },
            { name: "level", label: "Level" },
            { name: "opening_hours", label: "Opening Hours" },
            { name: "website", label: "Website" },
            { name: "phone", label: "Phone Number" }
        ]
});
nodeMarker.on('nodemarker:add',()=>{
	// because of async promise
	map.flyTo(nodeMarker.getLatLng())
})
nodeMarker.addTo(map);
```



And I suggest you to use overpass: [overpass-api](https://overpass-api.de/api/interpreter?data=%5Bout%3Ajson%5D%5Btimeout%3A25%5D%3B%0A%28%0A%20%20node%283060887211%29%3B%0A%29%3B%0Aout%20body%3B)
http://overpass-turbo.eu/s/1I3k

**Pull request closed without merging**